### PR TITLE
Backport ListBuffer.+= optimizations

### PR DESCRIPTION
**Pull request by retronym**
I'm seeing a 1.4x speedup for:

 ```
  @Benchmark
  public Object scalaListBufferPlusEq_212() {
    ListBuffer<String> buffer = new ListBuffer<>();
    int i = 0;
    while (i < size) {
      buffer.$plus$eq("");
      i += 1;
    }
    return buffer.result();
  }
```

2.12.8

```
[info] Benchmark                                 (size)   Mode  Cnt         Score        Error  Units
[info] ListsBenchmark.scalaListBufferPlusEq_212      10  thrpt    5  25856046.731 ± 1229100.335  ops/s
```

This patch:

```
[info] Benchmark                                 (size)   Mode  Cnt         Score        Error  Units
[info] ListsBenchmark.scalaListBufferPlusEq_212      10  thrpt    5  35848876.003 ± 514044.717  ops/s
```

It is still a little short of the 2.13.x performance; in which I saw:

```
[info] ListsBenchmark.scalaListBufferPlusEq          10  thrpt    5  37174742.519 ± 1304768.628  ops/s
[info] ListsBenchmark.scalaListBufferAddOne          10  thrpt    5  37201063.905 ± 2167146.358  ops/s
```

* in 2.13.x, the `scalaListBufferPlusEq` result requires `-XX:MaxInlineLevel=18`,
discussion at https://github.com/scala/bug/issues/11627)

### CODE REVIEW

**Review by dwijnand — APPROVED** [review]
I'm guessing the speedup is from not going through the accessors and, probably more importantly, not going through all the class-hierarchy indirection for `isEmpty`?

LGTM

**Comment by retronym** [comment]
The JIT compiled machine code for the benchmark in before and after this change is fully inlined. Note that the only x86 `call` instructions in the output of `jmh:run ... -prof perfasm` are to `Blackhole.consume` (benchmarking infrastructure to prevent dead code elimination of the code we are measuring).

https://gist.github.com/retronym/3dc4d3c1918c4f4ec46f958326484fc3

So the difference must lie elsewhere.

**Review by diesalbla — APPROVED** [review]

File: src/library/scala/collection/mutable/ListBuffer.scala [hunk 1]
```diff
@@ -171,15 +171,10 @@ final class ListBuffer[A]
...
-      last0 = new :: (x, Nil)
-      start = last0
-    } else {
-      val last1 = last0
-      last0 = new :: (x, Nil)
-      last1.tl = last0
-    }
+    ensureUnaliased()
+    val last1 = new ::[A](x, Nil)
+    if (len == 0) start = last1 else last0.tl = last1
```
Comment (diesalbla) [inline src/library/scala/collection/mutable/ListBuffer.scala:176]: Could the cause of speedup be that the code executes the allocation `new` and the rest of the code in parallel? In other words, now that the `new` is no longer control-dependent on the `if-then-else`?

**Review by eed3si9n — APPROVED** [review]
In terms of the correctness, the change looks good to me.

`ensureUnaliased()` just adds one indirection to `if (exported) copy()` (this encourages inlining?),

and `last0` would eventually get the same value as before.

File: src/library/scala/collection/mutable/ListBuffer.scala (hunk 1 above)
Comment (retronym) [reply src/library/scala/collection/mutable/ListBuffer.scala:176]: I'd rather not guess.

There is a way to systematically find out, but I'm not good enough at using the tools to do this without burning lots of hours.

## Whitebox introspection of the JIT compiler

To reduce the extraneous noise in the output, you'd first need to strip down the benchmark even further to reduce the amount of code to a bare minumum.

  - [Install](https://builds.shipilev.net/openjdk-jdk11/) or locally build OpenJDK with the `fastdebug` configuration.
  - Configure the C2 JIT compiler to dump diagnostics.  `bench/jmh:run bench.ListsBenchmark.scalaListBufferPlusEq_212  -f1 -wi 4 -i4  -psize=10 -jvm /code/jdk/build/macosx-x86_64-server-fastdebug/images/jdk/bin/java -jvmArgs -XX:+PrintIdeal -jvmArgs -XX:PrintIdealGraphLevel=2 -jvmArgs -XX:PrintIdealGraphFile=/tmp/graph.xml`
  - Inspect the compiler optimizations with the [Ideal Graph Visualizer](https://github.com/openjdk/jdk/tree/master/src/utils/IdealGraphVisualizer)
  - Compare fast and slow versions above.

## Blackbox comparison of the generated assembly

Someone used to reading x86 assembly (in particular that generated by HotSpot) could likely just glance at https://gist.github.com/retronym/3dc4d3c1918c4f4ec46f958326484fc3 and explain why the fast version is faster.

**Pull request merged**